Crafting day project - Shrinking header widget (web)

IIsaku opened the pull request:
## Checklist
- Contains unit tests ✅ 
- Contains breaking changes ❌
- Contains Atlas changes ❌
- Compatible with: MX 8️⃣, 9️⃣

#### Web specific
- Contains e2e tests ❌
- Is accessible ✅ 
- Compatible with Studio ✅ 
- Cross-browser compatible ✅ 

## This PR contains
- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [ ] Documentation
- [ ] Other (describe)

## What is the purpose of this PR?
Introduce a new widget to create a fixed header that shrinks when scrolling down a page.

## Relevant changes
- Shrink a sticky page header at a configured scroll threshold.
- Shrink a sticky page header while scrolling.

## What should be covered while testing?
Test all features on all supported browsers.

IIsaku committed:
feat: add shrinking header widget base for web

IIsaku committed:
feat(shrinking-header): use containment for scrollable area

IIsaku committed:
refactor(shrinking-header): change content area to div

- Change to div to also allow elements like a footer element to be added inside

IIsaku committed:
feat(shrinking-header): include functional styling

IIsaku committed:
feat(shrinking-header): first shrink header after scrolling content

IIsaku committed:
chore(shrinking-header): move to new package folder

IIsaku committed:
chore(shrinking-header-web): rename package folder

IIsaku committed:
chore(shrinking-header-web): wip

IIsaku committed:
chore: wip

IIsaku committed:
chore: move

IIsaku committed:
chore: rename package name old shrinking header

IIsaku committed:
feat(shrinking-header-web): shrink header at threshold

IIsaku committed:
feat(shrinking-header-web): introduce throttle

- Introduce lodash-es for throttle function
- Throttle evaluating shrunk state to improve performance
- Move evaluation function inside useEffect

IIsaku committed:
fix(shrinking-header-web): onscroll subscription

IIsaku committed:
feat(shrinking-header-web): improve throttle

IIsaku committed:
feat(shrinking-header-web): use sticky header

IIsaku committed:
feat(shrinking-header-web): listen to window scroll

- This is necessary for iOS Safari, so that it shrinks the header and removes the footer of the app.
- Bonus: it simplifies the widget a lot.

IIsaku committed:
chore(shrinking-header-web): document todos

IIsaku committed:
chore(shrinking-header-web): document todos

IIsaku committed:
feat(shrinking-header-web): calculate wrapping div height automatically

IIsaku committed:
feat(shrinking-header-web): calculate wrapping div height automatically

- Everytime the header changes size, the height is updated to the highest height occurrence.
- Use resize observer API, which is supported in every MX supported browser, to detect resizes.
- Remove shrunk class name property.

IIsaku committed:
fix(shrinking-header-web): shrinking after navigation

- Shrinking header now shrinks when you navigated from a different page
by removing the overflow rules from ancestor dom elements.
- Fix dependency security issue.

IIsaku committed:
feat(shrinking-header-web): improve shrink trigger

- Set both leading and trailing to tru for throttle to always execute the last scroll position.

IIsaku committed:
feat(shrinking-header-web): remove scrollable content property

IIsaku committed:
feat(shrinking-header-web): shrink while scrolling

- Shrink header linear to scroll from an init height to a shrunk height.
- Add TODOS in readme file.

IIsaku committed:
refactor(shrinking-header-web): extract setting wrapping div height logic

IIsaku committed:
refactor(shrinking-header-web): convert useWrappingDivStyle hook to useWrappingDivHeight

IIsaku committed:
feat(shrinking-header-web): conditional widget config props

IIsaku committed:
chore(shrinking-header-web): setup unit tests

- Add testing library for testing React Hooks. The same library is used for datagrid-web.
- Exclude test files in tsconfig.json, so that
these files aren't included during widget
compilation.
- Add the repo root node_modules folder to
typeRoots prop, so that jest typings can be
recognized. The node_modules/@types typeRoot is
kept in case typings will be added to the widget
package node_module folder.
- Add test cases for useWrappingDivHeight.

IIsaku committed:
chore(shrinking-header-web): fix header element typing

IIsaku committed:
test(shrinking-header-web): useWrappingDivHeight

IIsaku committed:
chore(shrinking-header-web): add custom tsconfig for unit tests

IIsaku committed:
test(shrinking-header-web): add unit tests

- Refactor to one tsconfig, because IDE's support only one in a directory.
- Extract Resize observer mocking.

IIsaku committed:
fix(shrinking-header-web): negative scroll behavior for iOS

IIsaku committed:
chore(shrinking-header-web): remove e2e test script

IIsaku committed:
chore: remove old widget implementation

Keraito commented on packages/pluggableWidgets/shrinking-header-web/package.json:
```diff
@@ -0,0 +1,38 @@
+  "widgetName": "ShrinkingHeader",
+  "version": "1.0.0",
+  "description": "A fixed header that shrinks when scrolling down a page",
+  "copyright": "Mendix B.V.",
```
You should probably change this copyright 😄

IIsaku replied:
I changed it a bit, but it's still property of Mendix, since I developed it during my crafting days. The only thing that differs is that it's community support.

Keraito commented on packages/tools/pluggable-widgets-tools/test-config/jest.config.js:
```diff
@@ -1,27 +1,42 @@
+        globals: {
+            "ts-jest": {
+                tsconfig: (await exists(join(projectDir, "tsconfig.spec.json")))
+                    ? join(projectDir, "tsconfig.spec.json")
```
Doesn't this always overwrite the config with `tsconfig.spec.json` if there is one? Is that what you want?

IIsaku replied:
Removed this code.

Keraito commented on packages/pluggableWidgets/shrinking-header-web/src/ShrinkingHeader.tsx:
```diff
@@ -0,0 +1,83 @@
+        [setShrinkingHeaderThresholdElement]
+    );
+
+    /*
+        Position "sticky" behaviour breaks when navigating from another page in the Mendix application.
+        This is due to the css rule `overflow-x: hidden` which is applied to `div class="mx-placeholder" />`.
+        This useEffect iterates through all ancestor elements and removes this rule, except for the body element.
+     */
```
Is it not possible to fix this through a CSS rule rather than through JS? Something along the lines of 
```
.mx-placeholder > .mx-your-shrinking-header-selector {
  overflow-x: visible;
}
```
or without the child selector?

IIsaku replied:
An `overflow-x: hidden` rule on any parent element except the body causes sticky positioned elements to break. This code is pretty limited in that it only removes inline styles.
To improve this situation I will add a note about the overflow issue in the upcoming documentation, so that users can anticipate on that. Besides that, I will simply resolve the `mx-placeholder` issue by adding the following to the widget styles instead of the code solution right now:
```
.mx-placeholder {
  overflow-x: visible !important;
}
```

Keraito commented on packages/pluggableWidgets/shrinking-header-web/src/ShrinkingHeader.tsx:
```diff
@@ -0,0 +1,83 @@
+    if (threshold) {
+        return (
+            <ShrinkingHeaderThreshold
+                rootElementRef={setShrinkingHeaderThreshold}
+                name={name}
+                tabIndex={tabIndex}
+                style={style}
+                className={className}
+                content={content}
```
Could be nice to extract these shared props to a const, would make it more readable and maintainable in the long run

IIsaku replied:
Good idea! Done.

Keraito commented on packages/pluggableWidgets/shrinking-header-web/src/components/ShrinkingHeaderLinear.tsx:
```diff
@@ -0,0 +1,65 @@
+    );
+
+    useEffect(() => {
+        const updateHeaderHeight = function (): void {
+            let headerHeight;
+
+            if (window.scrollY > initHeight - shrunkHeight) {
+                headerHeight = initHeight - (initHeight - shrunkHeight);
+            } else if (window.scrollY < 0) {
+                headerHeight = initHeight;
+            } else {
+                headerHeight = initHeight - window.scrollY;
+            }
+
+            setHeaderHeight(headerHeight);
+        };
```
Is there a reason you did not throttle this one compared to the ShrinkingHeaderThreshold one?

IIsaku replied:
Linear should act immediately, otherwise the animation isn't smooth. Throttling the threshold improved performance on mobile devices. In the future, we can provide a property so that users can configure it.

Keraito reviewed (commented):
Most of this looks fine, but if possible it would prob be nice to not traverse an entire DOM tree vertically to adjust a CSS rule through JS. I think it should be possible through CSS selectors, which would also be more efficient and performant. Please take a look whether you can address that in a different way.

IIsaku committed:
chore(shrinking-header-web): update copyright and repo url

IIsaku committed:
chore(shrinking-header-web): fix jest config & lock packages

IIsaku committed:
feat(shrinking-header-web): override mx-placeholder style with CSS

- Override mx-placeholder overflow-x hidden through CSS to not limit it to inline styles only.
- Refactor root component to contain common props in one constant.

IIsaku committed:
fix(shrinking-header-web): remove name prop usage

- Name prop isn't guaranteed to be unique. Also the name is applied as a className to the widget,
 which makes this prop unnecessary.

IIsaku committed:
feat(shrinking-header-web): improve widget properties

IIsaku committed:
chore(shrinking-header-web): make mx-placeholder comment more detailed

IIsaku committed:
feat(shrinking-header-web): improve component rendering

IIsaku committed:
feat(shrinking-header-web): introduce expressions for linear header

IIsaku committed:
fix(shrinking-header-web): update wrapping div height

- For header linear update wrapping div height when init height changes

IIsaku committed:
fix(shrinking-header-web): elements not clickable underneath wrapping div

IIsaku committed:
chore(shrinking-header-web): make code improvements

- Remove unused rootElementRef code.
- Simplify linear wrapping div height code.

IIsaku committed:
chore(shrinking-header-web): add documentation

- Add Changelog.md
- Improve README.md
- Improve widget property descriptions.

IIsaku committed:
chore(shrinking-header-web): add test project
